src/poseidon/matrix.rs

```diff
@@ -0,0 +1,675 @@
+//! acknowledgement: adapted from FileCoin Project: https://github.com/filecoin-project/neptune/blob/master/src/matrix.rs
+use std::iter::FromIterator;
+use std::ops::{Index, IndexMut};
+
+use ark_ff::PrimeField;
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Matrix<T: Clone>(pub Vec<Vec<T>>);
+
+impl<T: Clone> From<Vec<Vec<T>>> for Matrix<T> {
+    fn from(v: Vec<Vec<T>>) -> Self {
+        Matrix(v)
+    }
+}
+
+impl<T: Clone> Matrix<T> {
+    pub fn num_rows(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn num_columns(&self) -> usize {
+        if self.0.is_empty() {
+            0
+        } else {
+            let column_length = self.0[0].len();
+            for row in &self.0 {
+                if row.len() != column_length {
+                    panic!("not a matrix");
+                }
+            }
+            column_length
+        }
+    }
+
+    pub fn iter_rows<'a>(&'a self) -> impl Iterator<Item = &'a Vec<T>> {
+        self.0.iter()
+    }
+
+    pub fn is_square(&self) -> bool {
+        self.num_rows() == self.num_columns()
+    }
+
+    pub fn transpose(&self) -> Matrix<T> {
+        let size = self.num_rows();
+        let mut new = Vec::with_capacity(size);
+        for j in 0..size {
+            let mut row = Vec::with_capacity(size);
+            for i in 0..size {
+                row.push(self.0[i][j].clone())
+            }
+            new.push(row);
+        }
+        Matrix(new)
+    }
+}
+
+impl<T: Clone> Index<usize> for Matrix<T> {
+    type Output = Vec<T>;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+impl<T: Clone> IndexMut<usize> for Matrix<T> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.0[index]
+    }
+}
+
+// from iterator rows
+impl<F: Clone> FromIterator<Vec<F>> for Matrix<F> {
+    fn from_iter<T: IntoIterator<Item=Vec<F>>>(iter: T) -> Self {
+        let rows = iter.into_iter().collect::<Vec<_>>();
+        Self(rows)
+    }
+}
+
+impl<F: PrimeField> Matrix<F> {
+    /// return an identity matrix of size `n*n`
+    pub fn identity(n: usize) -> Matrix<F> {
+        let mut m = Matrix(vec![vec![F::zero(); n]; n]);
+        for i in 0..n {
+            m.0[i][i] = F::one();
+        }
+        m
+    }
+
+    pub fn is_identity(&self) -> bool {
+        if !self.is_square() {
```

Comment:
Aren't all these matrices square? The transpose method for example seems to assume so. Do we need non-square matrices anywhere?

Reply:
yes all MDS matrices should be square

Reply:
Then, we should notate this and make sure that we are using it everywhere in the interface and the implementation.